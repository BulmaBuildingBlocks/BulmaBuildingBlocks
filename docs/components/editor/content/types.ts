export enum EditableContentTypes {
  Text = 'text',
  Title = 'title'
}

export interface EditableContentProps {
  /**
   * Element wrapping the Editable Content.
   */
  tag?: string;

  /**
   * Provides different editing options
   */
  type?: EditableContentTypes;

  /**
   * Class placed on the parent element.
   */
  class?: string;

  /**
   * Value that's edited.
   */
  value: string;

  /**
   * Content Editable.
   */
  editable?: boolean;
}
