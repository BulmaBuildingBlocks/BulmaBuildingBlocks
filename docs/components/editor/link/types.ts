import { Status } from '~/shared/config';

export enum LinkType {
  Link = 'Link',
  Button = 'button',
  Rounded = 'button is-rounded',
  Outlined = 'button is-outlined',
  'Rounded Outlined' = 'button is-rounded is-outlined'
}

export class EditableLink {
  constructor(props: EditableLinkProps) {
    this.label = props.label;
    this.status = props.status || Status.None;
    this.linkType = props.linkType;
    this.href = props.href;
    this.popupPlacement = props.popupPlacement;
  }

  /*
   * Label on the button.
   */
  label: string;

  /**
   * Assigned hyperlink.
   */
  href?: string;

  /**
   * Link styles.
   */
  linkType: LinkType;

  /**
   * Link styles.
   */
  status: Status;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}

export interface EditableLinkProps {
  /**
   * Label on the button.
   */
  label: string;

  /**
   * Assigned hyperlink.
   */
  href?: string;

  /**
   * Link styles.
   */
  linkType: LinkType;

  /**
   * Link styles.
   */
  status?: Status;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}
