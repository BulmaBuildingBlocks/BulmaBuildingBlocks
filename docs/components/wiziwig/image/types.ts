export interface EditableImageProps {
  /**
   * Key of a permalink referencing the source of the picture.
   */
  src: string;

  /**
   * Caption on the picture, used also as alternative text.
   */
  alt: string;

  /**
   * Hyperlink attached to the picture.
   */
  href?: string;

  /**
   * Picture width.
   */
  width?: number;

  /**
   * Picture height.
   */
  height?: number;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}
