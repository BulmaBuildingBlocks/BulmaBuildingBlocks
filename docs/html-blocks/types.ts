import { VueConstructor } from 'vue';

export enum ContentTypes {
  Headers = 'headers',
  Features = 'features',
  CallToActions = 'calltoactions',
  Contents = 'contents',
  Footers = 'footers'
}

export interface Block {
  title: string;
  type?: ContentTypes;
  componentName: string;
  component: VueConstructor;
  content?: any;
  color?: string;
  container?: boolean;
}

export interface BlockListObject {
  [x: string]: VueConstructor;
}
