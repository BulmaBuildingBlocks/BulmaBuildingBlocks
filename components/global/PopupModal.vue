<template>
  <span class="is-relative">
    <transition name="fade">
      <span v-show="show" class="popper">
        <slot />
      </span>
    </transition>
    <slot name="reference" />
  </span>
</template>

<script lang="ts">
import { createPopper, Instance, Options } from '@popperjs/core';
import { Component, Prop, Vue, Watch } from 'nuxt-property-decorator';
import { VNode } from 'vue';

@Component
export default class PopupModal extends Vue {
  @Prop({ type: Object, default: {} }) options!: object;
  @Prop({ default: () => false }) show!: boolean;

  popperInstance?: Instance = undefined;
  popperOptions: Options = {
    strategy: 'absolute',
    placement: 'top',
    modifiers: [
      {
        name: 'offset',
        options: {
          offset: [0, 10]
        }
      },
      {
        name: 'preventOverflow',
        options: {
          padding: 5
        }
      },
      {
        name: 'flip',
        options: {
          fallbackPlacements: ['bottom']
        }
      },
      {
        name: 'computeStyles',
        options: {
          gpuAcceleration: false
        }
      }
    ]
  };

  $slots!: {
    reference: VNode[];
    default: VNode[];
  };

  @Watch('show')
  async onShowPopperChange() {
    if (this.popperInstance) {
      await this.popperInstance.update();
    }
  }

  created() {
    this.popperOptions = Object.assign(this.popperOptions, this.options);
  }

  mounted() {
    const referenceElm = this.$slots.reference[0].elm as HTMLElement;
    const popper = this.$slots.default[0].elm as HTMLElement;

    this.popperInstance = createPopper(
      referenceElm,
      popper,
      this.popperOptions
    );
  }

  destroy() {
    if (this.popperInstance) {
      this.popperInstance.destroy();
      this.popperInstance = undefined;
    }
  }
}
</script>
