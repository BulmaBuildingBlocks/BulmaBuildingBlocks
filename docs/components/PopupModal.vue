<template>
  <span>
    <span v-if="show" class="popup">
      <slot />
    </span>
    <slot name="reference" />
  </span>
</template>

<script lang="ts">
import { createPopper, Instance, Options } from '@popperjs/core';
import { Component, Prop, Vue, Watch } from 'nuxt-property-decorator';
import { VNode } from 'vue';

@Component
export default class PopupModal extends Vue {
  @Prop({ type: Object, default: {} }) options!: object;
  @Prop({ default: () => false }) show!: boolean;

  popperInstance?: Instance = undefined;
  popperOptions: Options = {
    strategy: 'absolute',
    placement: 'top',
    modifiers: [
      {
        name: 'offset',
        options: {
          offset: [0, 10]
        }
      },
      {
        name: 'preventOverflow',
        options: {
          padding: 5
        }
      },
      {
        name: 'flip',
        options: {
          fallbackPlacements: ['bottom', 'right', 'left']
        }
      },
      {
        name: 'computeStyles',
        options: {
          gpuAcceleration: false
        }
      }
    ]
  };

  referenceElm?: HTMLElement = undefined;
  content?: HTMLElement = undefined;

  $scopedSlots!: {
    reference: () => VNode[];
    default: () => VNode[];
  };

  @Watch('show', { immediate: true })
  async onShowPopperChange(show: boolean) {
    if (show) {
      await this.$nextTick();
      this.showPopup();
    } else {
      this.hidePopup();
    }
  }

  created() {
    this.popperOptions = Object.assign(this.popperOptions, this.options);
  }

  async showPopup() {
    this.referenceElm = (await this.$scopedSlots.reference()[0]
      .elm) as HTMLElement;
    this.content = (await this.$scopedSlots.default()[0].elm) as HTMLElement;

    if (this.referenceElm && this.content) {
      this.popperInstance = createPopper(
        this.referenceElm,
        this.content,
        this.popperOptions
      );

      await this.popperInstance.update();
    }
  }

  hidePopup() {
    if (this.popperInstance) {
      this.popperInstance.destroy();
      this.popperInstance = undefined;
    }
  }
}
</script>
