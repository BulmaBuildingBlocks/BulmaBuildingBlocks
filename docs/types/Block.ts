import { VueConstructor } from 'vue';

export interface Block {
  title: string;
  component: VueConstructor;
  content?: any;
  color?: string;
  container?: boolean;
}

export interface BlockListObject {
  [x: string]: VueConstructor;
}
