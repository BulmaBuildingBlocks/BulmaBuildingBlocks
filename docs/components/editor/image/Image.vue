<script lang="ts">
import { CreateElement } from 'vue';
import { Vue, Component, Prop, PropSync } from 'nuxt-property-decorator';
import ImageToolbar from './ImageToolbar.vue';
import PopupModal from '~/components/PopupModal.vue';

import { EditableImageProps } from '~/components/editor/image/types';
import ClickOutside from '~/directives/click-outside';

/***
 * Editting Features:
 * Upload and replace image
 * Alt tag
 * Link
 */

@Component({
  directives: {
    ClickOutside
  }
})
export default class EditableImage extends Vue {
  @Prop(Boolean) editable!: boolean;
  @PropSync('value', { type: Object }) newValue!: EditableImageProps;

  modalOpen = false;

  toggleModal(event: Event) {
    event.preventDefault();

    this.modalOpen = !this.modalOpen;
  }

  closeModal() {
    this.modalOpen = false;
  }

  render(createElement: CreateElement) {
    const image = createElement('img', {
      attrs: {
        src: this.newValue.src,
        alt: this.newValue.alt,
        width: this.newValue.width,
        height: this.newValue.height
      },
      on: {
        click: this.toggleModal
      }
    });

    const imageWithLink = createElement(
      'a',
      {
        attrs: {
          href: this.newValue.href
        }
      },
      [image]
    );

    const imageToolbar = createElement(ImageToolbar, {
      props: {
        value: this.newValue
      },
      on: {
        closeModal: () => {
          this.closeModal();
        },
        'update:value': (value: EditableImageProps) => {
          this.newValue = value;
        }
      }
    });

    const popupModal = createElement(PopupModal, {
      scopedSlots: {
        reference: () => {
          return this.newValue.href ? imageWithLink : image;
        },
        default: () => {
          return imageToolbar;
        }
      },
      directives: [
        {
          name: 'click-outside',
          value: { handler: this.closeModal }
        }
      ],
      props: {
        show: this.editable && this.modalOpen,
        options: {
          placement: this.newValue.popupPlacement || 'right'
        },
        modifiers: [
          {
            name: 'flip',
            options: {
              fallbackPlacements: ['left', 'top', 'right', 'bottom']
            }
          }
        ]
      }
    });

    return popupModal;
  }
}
</script>
