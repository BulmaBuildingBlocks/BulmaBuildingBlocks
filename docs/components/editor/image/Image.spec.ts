import { mount, Wrapper } from '@vue/test-utils';
import Image from '~/components/editor/image/Image.vue';
import { EditableImageProps } from '~/components/editor/image/types';
import PopupModal from '~/components/PopupModal.vue';
import ClickOutside from '~/directives/click-outside';
import ImageToolbar from '~/components/editor/image/ImageToolbar.vue';

const createWrapper = (propsData: EditableImageProps) => {
  return mount(Image, {
    propsData: { value: propsData, editable: true },
    stubs: {
      PopupModal,
      ImageToolbar
    },
    directives: {
      ClickOutside
    }
  });
};

describe('Image.vue', () => {
  let wrapper: Wrapper<Image & { [key: string]: any }>;

  beforeEach(() => {
    wrapper = createWrapper({
      alt: 'Stock Image',
      src: './icon.png'
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('has the expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('has the expected html structure with link', () => {
    wrapper = createWrapper({
      alt: 'Click Here',
      src: './click-here.png',
      href: 'http://www.bulmabuildingblocks.com/'
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy();
  });

  test('toggle model toggles correctly', () => {
    wrapper.vm.toggleModal(new Event('clicked'));
    expect(wrapper.vm.modalOpen).toEqual(true);
    wrapper.vm.toggleModal(new Event('clicked'));
    expect(wrapper.vm.modalOpen).toEqual(false);
  });

  test('close modal function', () => {
    expect(typeof wrapper.vm.closeModal).toBe('function');

    wrapper.vm.closeModal();
    expect(wrapper.vm.modalOpen).toEqual(false);
  });

  test('open and close popup toolbar', async () => {
    // Check button and check popup is shown
    await wrapper.find('img').trigger('click');
    expect(wrapper.find('.popup').exists()).toBeTruthy();

    // Check button and check popup is hidden
    await wrapper.find('.card-header-icon').trigger('click');

    expect(wrapper.find('.popup').exists()).toBeFalsy();
    // Check button and check popup is shown
    await wrapper.find('img').trigger('click');

    expect(wrapper.find('.popup').exists()).toBeTruthy();
    // Check button and check popup is hidden
    await wrapper.find('img').trigger('click');
    expect(wrapper.find('.popup').exists()).toBeFalsy();
  });
});
