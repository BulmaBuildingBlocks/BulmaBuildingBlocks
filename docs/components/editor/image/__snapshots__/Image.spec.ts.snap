// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image.vue has the expected html structure 1`] = `
<span>
  <!---->
   
  <img
    alt="Stock Image"
    src="./icon.png"
  />
</span>
`;

exports[`Image.vue has the expected html structure with link 1`] = `
<span>
  <!---->
   
  <a
    href="http://www.bulmabuildingblocks.com/"
  >
    <img
      alt="Click Here"
      src="./click-here.png"
    />
  </a>
</span>
`;
